synadm/api.py

```diff
@@ -304,15 +304,11 @@ def generate_mxid(self, user_id):
         """
         if user_id is None:
             return None
-        elif any(x in user_id for x in ["@", ":"]):
-            if re.match(r"\@.+\:.+", user_id):
-                return user_id
-            else:
-                mxid = "@{}:{}".format(re.sub("[@:]", "", user_id),
-                                       self.server_name())
-                return mxid
+        elif re.match(r"@[\w\d\_\.\_\=\-\/]+:[\w\d]+", user_id):
```

Comment:
In the localpart you have 2 times the underscore. Also some of it don't need to be escaped. I'd suggest this one
`@[-_/=\w\d\.]+`

In the servername part I think we should _at least_ allow everything that's allowed in http urls. So the word-charachters (\w) don't include the dash and the dot. We need to add it. I just realize that I couldn't tell by heart if there's anything else allowed....these days...

Reply:
Perhaps we can use more information of the server name format as defined in the matrix specification, which can be found here: https://spec.matrix.org/v1.2/appendices/#server-name. 
Let me know if you think we should rely at least in part on the server name specification or if we should only use "normal" domain format for the validation.

Reply:
Fixed the regex and added domain name validation in 9dfb4652712e3ac67549920ab3e81c304d35d371